### PR TITLE
nixos/shells-environment: allow int and float in environment variables

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/config/shells-environment.nix [hunk 1]
```diff
@@ -42,8 +42,8 @@ in
         strings.  The latter is concatenated, interspersed with colon
         characters.
       '';
-      type = with types; attrsOf (oneOf [ (listOf str) str path ]);
-      apply = mapAttrs (n: v: if isList v then concatStringsSep ":" v else "${v}");
+      type = with types; attrsOf (oneOf [ (listOf (oneOf [ float int str ])) float int str path ]);
+      apply = mapAttrs (n: v: if isList v then concatStringsSep ":" v else toString v);
```
Comment (NickCao) [inline nixos/modules/config/shells-environment.nix:46]: `concatStringSep` cannot handle a list of float or int.

Comment (SuperSandro2000) [reply nixos/modules/config/shells-environment.nix:46]: I must admit, I didn't test that explicitly, only plain values :/

Comment (SuperSandro2000) [reply nixos/modules/config/shells-environment.nix]: I've added a map to fix that.

Comment (ambroisie) [reply nixos/modules/config/shells-environment.nix]: `concatMapStringsSep`?

Comment (SuperSandro2000) [reply nixos/modules/config/shells-environment.nix]: I don't get your question

Comment (NickCao) [reply nixos/modules/config/shells-environment.nix]: > I don't get your question

`lib.concatMapStringsSep ":" toString [ 1 2 3 true false "test" ]`

File: nixos/modules/config/shells-environment.nix [hunk 2]
```diff
@@ -42,8 +42,8 @@ in
         strings.  The latter is concatenated, interspersed with colon
         characters.
       '';
-      type = with types; attrsOf (oneOf [ (listOf str) str path ]);
-      apply = mapAttrs (n: v: if isList v then concatStringsSep ":" v else "${v}");
+      type = with types; attrsOf (oneOf [ (listOf (oneOf [ float int str ])) float int str path ]);
+      apply = mapAttrs (n: v: if isList v then concatStringsSep ":" (map toString v) else toString v);
```
Comment (NickCao) [inline nixos/modules/config/shells-environment.nix:46]:
```suggestion
      apply = mapAttrs (n: v: if isList v then concatMapStringsSep ":" toString v else toString v);
```

**Review by NickCao — COMMENTED** [review]
Minor simplification, otherwise LGTM.

**Review by ambroisie — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Do we want to merge this?

**Comment by NickCao** [comment]
> Do we want to merge this?

Of course!

**Review by NickCao — APPROVED** [review]

**Comment by PerchunPak** [comment]
Just a little note, this might be a breaking change, because:
```nix
nix-repl> "${/etc/nixos}"
"/nix/store/92iarj2qa7kfalx247sqa333i995zycr-nixos"

nix-repl> toString /etc/nixos
"/etc/nixos"
```

See https://github.com/NixOS/nix/issues/7327

**Comment by jtojnar** [comment]
Not a fan of automatic coercions but I guess for `int`s the only downside is increased complexity.

We should not stringify floats, though, because it is lossy in Nix and can be pretty confusing https://github.com/NixOS/nix/issues/5733#issuecomment-1529002378

**Comment by SuperSandro2000** [comment]
please see #308779

**Pull request merged**